Library/SPIClient.m

```diff
@@ -1380,14 +1379,10 @@ - (void)onSpiConnectionStatusChanged:(SPIConnectionState)newConnectionState {
  */
 - (void)startPeriodicPing {
     __weak __typeof(&*self) weakSelf = self;
-    
-    if (_pingTimer != nil) {
-        [_pingTimer cancel];
-        _pingTimer = nil;
-    }
-    _pingTimer = [[SPIGCDTimer alloc] initWithObject:self queue:"com.assemblypayments.ping"];
-    
-    [_pingTimer afterDelay:0 repeat:true block:^(id self) {
+
+    _pingTimer = [[SPIRepeatingTimer alloc] initWithQueue:"com.assemblypayments.ping"
+                                                 interval:pingFrequency
+                                                    block:^{
         if (!weakSelf.connection.isConnected || weakSelf.secrets == nil) {
             dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
                 [weakSelf stopPeriodPing];
```

Comment:
we are nullifying the _pingTimer itself (by calling stopPeriodPing) while the work is still happening. this should not matter, right?

Reply:
That's fine, GCD queues are retained by any currently-running blocks, so if the SPIRepeatingTimer is nilled, the queue keep running until that block completes, and then clean up tidily.